crates/shared/src/price_estimation/quasimodo.rs

```diff
@@ -0,0 +1,350 @@
+use crate::bad_token::BadTokenDetecting;
+use crate::baseline_solver::BaseTokens;
+use crate::http_solver_api::model::{
+    AmmModel, AmmParameters, BatchAuctionModel, ConstantProductPoolParameters, CostModel, FeeModel,
+    OrderModel, TokenInfoModel,
+};
+use crate::http_solver_api::HttpSolverApi;
+use crate::price_estimation::gas::{
+    ERC20_TRANSFER, GAS_PER_ORDER, GAS_PER_UNISWAP, INITIALIZATION_COST, SETTLEMENT,
+};
+use crate::price_estimation::{
+    ensure_token_supported, Estimate, PriceEstimating, PriceEstimationError, Query,
+};
+use crate::recent_block_cache::Block;
+use crate::sources::uniswap::pool_cache::PoolCache;
+use crate::sources::uniswap::pool_fetching::PoolFetching;
+use crate::token_info::TokenInfoFetching;
+use ethcontract::{H160, U256};
+use gas_estimation::GasPriceEstimating;
+use model::order::OrderKind;
+use model::TokenPair;
+use num::{BigInt, BigRational};
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+pub struct QuasimodoPriceEstimator {
+    pub api: Arc<HttpSolverApi>,
+    pub pools: Arc<PoolCache>,
+    pub bad_token_detector: Arc<dyn BadTokenDetecting>,
+    pub token_info: Arc<dyn TokenInfoFetching>,
+    pub gas_info: Arc<dyn GasPriceEstimating>,
+    pub native_token: H160,
+    pub base_tokens: Arc<BaseTokens>,
+}
+
+impl QuasimodoPriceEstimator {
+    async fn estimate(&self, query: &Query) -> Result<Estimate, PriceEstimationError> {
+        if query.buy_token == query.sell_token {
+            return Ok(Estimate {
+                out_amount: query.in_amount,
+                gas: 0.into(),
+            });
+        }
+
+        ensure_token_supported(query.buy_token, self.bad_token_detector.as_ref()).await?;
+        ensure_token_supported(query.sell_token, self.bad_token_detector.as_ref()).await?;
+
+        let gas_cost = U256::from_f64_lossy(self.gas_info.estimate().await?.legacy);
```

Comment:
Just noticed this now, this might not work as expected for EIP-1559 estimates. I suggest converting the estimate into an effective gas price:

```suggestion
        let gas_cost = U256::from_f64_lossy(self.gas_info.estimate().await?.effective_gas_price());
```